src/ops/conversion.rs

```diff
@@ -0,0 +1,244 @@
+use mlx_macros::default_device;
+
+use crate::{Array, ArrayElement, StreamOrDevice};
+
+impl Array {
+    /// Cast the array to a specified type.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use mlx::Array;
+    ///
+    /// let array = Array::from_slice(&[1i16,2,3], &[3]);
+    /// let mut new_array = array.as_type::<f32>();
+    /// new_array.eval();
+    ///
+    /// assert_eq!(new_array.dtype(), mlx::Dtype::Float32);
+    /// assert_eq!(new_array.shape(), &[3]);
+    /// assert_eq!(new_array.item_size(), 4);
+    /// assert_eq!(new_array.as_slice::<f32>(), &[1.0,2.0,3.0]);
+    /// ```
+    #[default_device]
+    pub fn as_type_device<T: ArrayElement>(&self, stream: StreamOrDevice) -> Array {
+        unsafe {
+            let new_array = mlx_sys::mlx_astype(self.c_array, T::DTYPE as u32, stream.as_ptr());
+            Array::from_ptr(new_array)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use half::{bf16, f16};
+
+    use crate::complex64;
+
+    use super::*;
+
+    macro_rules! test_as_type {
+        ($src_type:ty, $src_val:expr, $dst_type:ty, $dst_val:expr, $len:expr) => {
+            paste::paste! {
+                #[test]
+                fn [<test_as_type_ $src_type _ $dst_type>]() {
+                    let array = Array::from_slice(&[$src_val; $len], &[$len as i32]);
+                    let s = StreamOrDevice::default();
+
+                    let mut new_array = array.as_type_device::<$dst_type>(s);
```

Comment:
not that it matters much but you should be able to leverage the default device method here. `as_type`

Reply:
Yeah, this is mostly generated by copilot, so I didn't bother changing that. But I agree that testing`as_type` is better here